lib/transport/ip/http-client.js

```diff
@@ -59,37 +59,70 @@ class HttpClient extends EventEmitter {
   }
 
   /**
-   * Attempt to pair with a device.
+   * Begins the pairing process. For devices with random pins, this
+   * will cause it to show the pin on the screen.
    *
-   * @param {string} pin - The pairing PIN
-   * @returns {Promise} Promise which resolves when pairing is complete.
+   * @returns {Promise} Promise which resolves to opaque
+   * pairing data when complete.
+   */
+  async startPairing() {
+    const connection = this._pairingConnection =
+      new HttpConnection(this.address, this.port);
+
+    // M1
+    const packet = await this.pairingProtocol.buildPairSetupM1();
+    const response = await connection.post('/pair-setup', packet,
+                                           'application/pairing+tlv8');
+
+    // M2
+    return this.pairingProtocol.parsePairSetupM2(response.body);
+  }
+
+  /**
+   * Finishes a pairing process that began with startPairing()
+   *
+   * @param {Object} pairingData - The pairing data returned from startPairing()
+   * @returns {Promise} Promise which resolvew when pairing is complete.
    */
-  pairSetup(pin) {
+  async finishPairing(pairingData, pin) {
+    if (!pairingData || !this._pairingConnection) {
+      throw new Error('Must call startPairing() first');
+    }
+
     const re = /^\d{3}-\d{2}-\d{3}$/;
     if (!re.test(pin)) {
       return Promise.reject('Invalid PIN');
     }
 
-    const connection = new HttpConnection(this.address, this.port);
-    return this.pairingProtocol.buildPairSetupM1().then((packet) => {
-      return connection.post('/pair-setup', packet, 'application/pairing+tlv8');
-    }).then((response) => {
-      return this.pairingProtocol.parsePairSetupM2(response.body);
-    }).then((tlv) => {
-      return this.pairingProtocol.buildPairSetupM3(tlv, pin);
-    }).then((packet) => {
-      return connection.post('/pair-setup', packet, 'application/pairing+tlv8');
-    }).then((response) => {
-      return this.pairingProtocol.parsePairSetupM4(response.body);
-    }).then(() => {
-      return this.pairingProtocol.buildPairSetupM5();
-    }).then((packet) => {
-      return connection.post('/pair-setup', packet, 'application/pairing+tlv8');
-    }).then((response) => {
-      return this.pairingProtocol.parsePairSetupM6(response.body);
-    }).then(() => {
-      connection.close();
-    });
+    const connection = this._pairingConnection;
```

Comment:
It would probably be wise to `delete this._pairingConnection` at this point.

Reply:
yup good catch